### ISSUE TITLE
gRPC JSON service

### ISSUE THREAD

**Issue by tokuhirom** [line/armeria#255]
I want a gRPC json service. such as `GrpcJsonServiceBuilder.service(myService).build()`.
On gRPC, there's no JSON support. as a result, it's hard to use on programming languages, that doesn't have a gRPC client library.

I know there's a [grpc-gateway](https://github.com/grpc-ecosystem/grpc-gateway). But if there's a GrpcJsonService, it may be advantage of armeria.

**Comment by trustin** [comment]
Sounds like a good idea. `GrpcService` currently delegates all the stream processing to gRPC, which means we cannot support other serialization formats such as Thrift and JSON at the current state. We will have to implement our own gRPC implementation layer as a service/client decorator as originally proposed by @anuraaga

**Comment by anuraaga** [comment]
Just wondering, where would you want a json version? If browser, protobuf has good support for binary in browsers now and it's the preferred way of using it over json now. I am currently working on a JavaScript library that will speak a slightly modified grpc protocol from the browser (trailing headers are sent as a data frame).

Grpc is just a wire protocol, and they officially support plugging in other serialization formats if you want, so actually us using the grpc library as is should get the support for free. Proto3 supports json encoding, haven't looked at details but it should be easy to plug in already.

The client would still need a library that speaks the grpc wire protocol. Grpc is just a wire protocol so not using it means not using grpc; for use from languages where you don't want to, a proxy like grpc gateway is required. It would be easy to implement such a proxy in armeria as a decorator of the grpc service though - my no-trailers mod would probably be similar.

**Comment by anuraaga** [comment]
On phone so can't edit post meant to include this link that documents the pluggability 

http://www.grpc.io/faq/

**Comment by eonezhang** [comment]
if there is a proxy layer can do auto protocol transformation would be great.

**Comment by anuraaga** [comment]
I'm currently working on heavy refactoring of the codebase but intend to support translation of JSON serialization soon as the work finishes up.

**Comment by eonezhang** [comment]
hi @anuraaga, after #499 was merged into master, now did Armeria support gRPC call by JSON other than protobuf?

**Comment by anuraaga** [comment]
Not yet - some JSON-related code is under code review, this PR for now allows binary proto requests to be made without grpc framing (just post the bytes). Most languages, including JavaScript, support binary proto, if not grpc, so I'd recommend seeing if that's an option before looking at JSON.

**Comment by trustin** [comment]
@anuraaga Is this issue resolved? /cc @tokuhirom

**Comment by anuraaga** [comment]
Ah, yeah it is :D
